Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,582 @@
+            {
+                ConfigureRequest(request);
+               
+                response = Client.SendAsync(request).Result;
+                if (!response.IsSuccessStatusCode)
+                    return null;
+                return JsonConvert.DeserializeObject<TokenResponse>(response.Content.ReadAsStringAsync().Result);
+                
+            }
+
+        }
+
+        private bool DecodeToken(TokenResponse token)
+        {
+            try
+            {
+                JObject tokenContent = DecodeJWT(token.id_token);
+
+                // Translate UserDetails according to Token
+                if (!tokenContent.TryGetValue("sub", out JToken sub))
+                    throw new Exception("Token data was not derived from valid JWT: missing 'sub'");
+                UserData userData = new UserData(sub.ToString());
+
+                userData.Initialize();
+
+                if (!tokenContent.TryGetValue("name", out JToken userName))
+                    throw new Exception("Token data was not derived from valid JWT: missing 'name'");
+                userData.Username = userName.ToString();
+
+                if (tokenContent.TryGetValue("given_name", out JToken firstName))
+                    userData.FirstName = firstName.ToString();
+                if (tokenContent.TryGetValue("family_name", out JToken lastName))
+                    userData.LastName = lastName.ToString();
+                if (tokenContent.TryGetValue("phone_number", out JToken phoneNumber))
+                    userData.PhoneNumber = phoneNumber.ToString();
+                if (tokenContent.TryGetValue("email", out JToken email))
+                    userData.EmailAddress = email.ToString();
+              
+                try
+                {
+                        userData.Roles = tokenContent.GetOrDefault(RolesClaim).ToObject<List<string>>();
+                }
+                catch (Exception ex)
+                {
+                    LastException = ex;
+                    Log.Publish(MessageLevel.Warning, MessageFlags.SecurityMessage, "DecodingTokenError", "Failed to decode Roles Claim.", "Failed to decode Claim for roles.", ex);
+                }
+
+                // Roles will need to be obtained from a Claim
+                userData.IsDefined = true;
+                userData.IsDisabled = false;
+                userData.IsLockedOut = false;
+                UserData = userData;
+                return true;
+            }
+            catch (Exception ex)
+            {
+                string message = $"Exception occurred while decoding the OpenID token: {ex.Message}";
+                throw new Exception(message, ex);
+            }
+        }
+
+        private JObject DecodeJWT(string token)
+        {
+            if (!token.Contains("."))
+                throw new InvalidExpressionException("A valid JQT token requires at least one '.'");
+
+            string jsoeHeader = token.Substring(0, token.IndexOf("."));
```

Comment:
Is it supposed to be `jsonHeader`?

Reply:
No it's JSOE. 

> 
   JOSE Header
      JSON object containing the parameters describing the cryptographic operations and parameters employed.  The JOSE (JSON Object Signing and Encryption) Header is comprised of a set of Header Parameters.

it is in Json format but this specific header is called JSOE.

Reply:
Gotcha. I'd never seen that acronym before.